pkg/controller/volume/expand/expand_controller.go

```diff
@@ -123,107 +113,146 @@ func NewExpandController(
-		expc.pvcLister,
-		expc.pvLister,
-		&expc.volumePluginMgr,
-		kubeClient)
 	return expc, nil
 }
 
-func (expc *expandController) Run(stopCh <-chan struct{}) {
-	defer runtime.HandleCrash()
-	klog.Infof("Starting expand controller")
-	defer klog.Infof("Shutting down expand controller")
+func (expc *expandController) enqueuePVC(obj interface{}) {
+	pvc := obj.(*v1.PersistentVolumeClaim)
+	size := pvc.Spec.Resources.Requests[v1.ResourceStorage]
+	statusSize := pvc.Status.Capacity[v1.ResourceStorage]
 
-	if !controller.WaitForCacheSync("expand", stopCh, expc.pvcsSynced, expc.pvSynced) {
-		return
+	if pvc.Status.Phase == v1.ClaimBound && size.Cmp(statusSize) > 0 {
+		key, err := kcache.DeletionHandlingMetaNamespaceKeyFunc(pvc)
+		if err != nil {
+			runtime.HandleError(fmt.Errorf("Couldn't get key for object %#v: %v", pvc, err))
+			return
+		}
+		expc.queue.Add(key)
 	}
+}
 
-	// Run volume sync work goroutine
-	go expc.syncResize.Run(stopCh)
-	// Start the pvc populator loop
-	go expc.pvcPopulator.Run(stopCh)
-	<-stopCh
+func (expc *expandController) processNextWorkItem() bool {
+	key, shutdown := expc.queue.Get()
+	if shutdown {
+		return false
+	}
+	defer expc.queue.Done(key)
+
+	err := expc.syncHandler(key.(string))
+	if err == nil {
+		expc.queue.Forget(key)
+		return true
+	}
+
+	runtime.HandleError(fmt.Errorf("%v failed with : %v", key, err))
+	expc.queue.AddRateLimited(key)
+
+	return true
 }
 
-func (expc *expandController) deletePVC(obj interface{}) {
-	pvc, ok := obj.(*v1.PersistentVolumeClaim)
-	if !ok {
-		tombstone, ok := obj.(kcache.DeletedFinalStateUnknown)
-		if !ok {
-			runtime.HandleError(fmt.Errorf("couldn't get object from tombstone %+v", obj))
-			return
-		}
-		pvc, ok = tombstone.Obj.(*v1.PersistentVolumeClaim)
-		if !ok {
-			runtime.HandleError(fmt.Errorf("tombstone contained object that is not a pvc %#v", obj))
-			return
+func (expc *expandController) syncHandler(key string) error {
+	namespace, name, err := kcache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	pvc, err := expc.pvcLister.PersistentVolumeClaims(namespace).Get(name)
+	if errors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		klog.V(5).Infof("Error getting PVC %q (uid: %q) from informer : %v", util.GetPersistentVolumeClaimQualifiedName(pvc), pvc.UID, err)
+		return err
+	}
+
+	pv, err := getPersistentVolume(pvc, expc.pvLister)
+	if err != nil {
+		klog.V(5).Infof("Error getting Persistent Volume for PVC %q (uid: %q) from informer : %v", util.GetPersistentVolumeClaimQualifiedName(pvc), pvc.UID, err)
+		return err
+	}
+	if pv.Spec.ClaimRef == nil || pvc.Namespace != pv.Spec.ClaimRef.Namespace || pvc.Name != pv.Spec.ClaimRef.Name {
+		err := fmt.Errorf("Persistent Volume is not bound to PVC being updated : %s", util.ClaimToClaimKey(pvc))
+		klog.V(4).Infof("%v", err)
+		return err
+	}
+
+	volumeSpec := volume.NewSpecFromPersistentVolume(pv, false)
+	volumePlugin, err := expc.volumePluginMgr.FindExpandablePluginBySpec(volumeSpec)
```

Comment:
hmm, yeah we could.  but shortly after this change, we are going to make changes to annotate the PVC when migration is enabled and PVC should be resized by external resizer. I was thinking it is better to do this in `expand_controller.go` rather than inside operation_generator. 

In worst case, this will cause plugin look up twice, which doesn't seem too expensive. wdyt?

Reply:
Performance-wise it's ok, I don't like big chunk of code almost copied on two places.

Reply:
Ok, I can see this part spits an event about external resizer and we probably don't want to have this piece in OperationGenerator because it would mess up with metrics. So leave it as it is.